arbor/util/unwind.cpp

```diff
@@ -1,124 +1,55 @@
-    unw_getcontext(&context);
-    unw_init_local(&cursor, &context);
-
-    while (unw_step(&cursor) > 0) {
-        // find the stack position
-        unw_word_t offset, pc;
-        unw_get_reg(&cursor, UNW_REG_IP, &pc);
-        if (pc == 0) {
-            break;
-        }
-
-        // get the name 
-        char sym[512];
-        if (unw_get_proc_name(&cursor, sym, sizeof(sym), &offset) == 0) {
-            frames_.push_back({std::string(sym), pc});
-        } else {
-            frames_.push_back({std::string("???"), pc});
-        }
+#ifdef WITH_BACKTRACE
+    auto bt = boost::stacktrace::basic_stacktrace{};
+    for (const auto& f: bt) {
+        frames_.push_back(source_location{f.name(), f.source_file(), f.source_line()});
     }
-}
-
-std::string demangle(std::string s) {
-    int status;
-    char* demangled = abi::__cxa_demangle(s.c_str(), nullptr, nullptr, &status);
-
-    // __cxa_demangle only returns a non-empty string if it is passed a valid C++
-    // mangled c++ symbol (i.e. it returns an empty string for normal c symbols)
-    if (status==0) {
-        s = demangled;
-    }
-    std::free(demangled); // don't leak the demangled string
-
-    return s;
+#endif
 }
 
 std::ostream& operator<<(std::ostream& out, const backtrace& trace) {
-    for (auto& f: trace.frames_) {
-        char loc_str[64];
-        snprintf(loc_str, sizeof(loc_str), "0x%lx", f.position);
-        out << loc_str << " " << f.name << "\n";
-        if (f.name=="main") {
-            break;
-        }
+#ifdef WITH_BACKTRACE
+    out << "Backtrace:\n";
+    int ix = 0;
+    for (const auto& f: trace.frames_) {
+        out << std::setw(8) << ix << " " << f.func << " (" << f.file << ":" << f.line << ")\n";
+        ix++;
     }
+#endif
     return out;
 }
 
-#if 0
-// Temporarily deprecated: automatic writing to disk of strack traces
-// needs to be run-time configurable.
-
-void backtrace::print(bool stop_at_main) const {
-    using namespace arb::memory::util;
-
-    auto i = 0;
-    while (file_exists("backtrace_" + std::to_string(i))) {
-        ++i;
-    }
-    auto fname = "backtrace_" + std::to_string(i);
-    auto fid = std::ofstream(fname);
-    for (auto& f: frames_) {
-        char loc_str[64];
-        snprintf(loc_str, sizeof(loc_str),"0x%lx", f.position);
-        fid << loc_str << " " << f.name << "\n";
-        if (stop_at_main && f.name=="main") {
-            break;
-        }
-    }
-    std::cerr << "BACKTRACE: A backtrace was generated and stored in the file " << fname << ".\n";
-    std::cerr << "           View a brief summary of the backtrace by running \"scripts/print_backtrace " << fname << " -b\".\n";
-    std::cerr << "           Run \"scripts/print_backtrace -h\" for more options.\n";
+backtrace& backtrace::pop(std::size_t n) {
+    auto end = frames_.begin();
+    for(std::size_t ix = 0;
+        ix < n && end != frames_.end();
+        ++ix, ++end) {}
+    frames_.erase(frames_.begin(), end);
```

Comment:
```suggestion
    frames_.erase(frames_.begin(), frames_.begin() + std::min(frames_.size(),n));
```

Reply:
That's what I changed it from? Like it better? Sure :D